### PR TITLE
Review9 3

### PR DESCRIPTION
**Pull request by woitech**

### CODE REVIEW

File: kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/AppConfig.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+// dummy implementation of config class. Normally it should read data from config file, for example config.xml
+public final class AppConfig {
+    private Company appUser;
+    private DemandRetriever demandRetriever;
+    private  StorageService storageService;
+    private  Map<Company, OfferService> offerServices;
+    private  Map<Company, OrderService> orderServices;
+    private OrderProcessorResultHandler orderProcessorResultHandler;
+
+    public AppConfig(String configFile) {
```
Comment (aszecowka) [inline kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/AppConfig.java]: If it is dummy config, do not pretend that data is read from configFile - it is misleading.

File: kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/AppConfig.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    private boolean parseConfigFile(String configFile) {
+        // dummy implementation
+
+        appUser = new Company(1, "Food2Door", "ul. Handlowa 10, 12-345 Fudowo", "food2door@example.com", "123456789");
+        storageService = new MySqlStorageService("jdbc:mysql://localhost:3306/storage", "supply_processor",
+                                                 "IUyhgTy^599*JH");
+        demandRetriever = new DefaultDemandRetriever();
+
+        Company efs = new Company(101, "ExtraFoodShop", "ul. Rolnicza 1, 11-111 Kapustowo", "efs@example.com",
```
Comment (aszecowka) [inline kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/AppConfig.java]: Let assume that I want to add new company and their offer - it means that I have to modify this method - this violate Open/Closed principle.

File: kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/process/OrderProcessor.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        }
+        this.config = config;
+    }
+
+    public OrderProcessorResult process(Demand demand) {
+        if (demand == null) {
+            throw new IllegalArgumentException();
+        }
+        Map<Company, Offer> offers = collectOffers(config.getOfferServices());
+        Rules rules = config.getRules();
```
Comment (aszecowka) [inline kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/process/OrderProcessor.java]: unused variable

File: kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/process/Rules.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package com.kodilla.good.patterns.food2door.process;
+
+public class Rules {
```
Comment (aszecowka) [inline kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/process/Rules.java]: You should focus on main purpose of this task: it was about enabling adding new Companies to the Food2Door offer - there is too much additional code (like Rules here). Keep it simple :)

File: kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/data/Order.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        if (date == null) {
+            throw new IllegalArgumentException("Date is null");
+        }
+        if (producer == null) {
+            throw new IllegalArgumentException("Producer is null");
+        }
+        if (recipient == null) {
+            throw new IllegalArgumentException("Recipient is null");
+        }
+        if (productName == null || productName.isEmpty()) {
```
Comment (aszecowka) [inline kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/data/Order.java]: You can add static method in helper class which will simplify string validation - as you can notice similar code is repeated in many places. Another option is to use library which implements JSR-303, for example http://hibernate.org/validator/

File: kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/data/Order.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    private final Company recipient;
+    private final String productName;
+    private final String measure;
+    private double quantity;
+    private final double price;
+    // derived
+    private final double amount;
+
+    public Order(final LocalDate  date, final Company producer, final Company recipient, final String productName,
+                 final String measure, final double quantity, final double price) {
```
Comment (aszecowka) [inline kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/data/Order.java]: this constructor has too many arguments. 
ProductName, measure, price - they can be group together in Product class. Product can also has reference to Producer - and thanks to that we can reduce number of arguments to 4.

**Review by aszecowka — CHANGES_REQUESTED** [review]

File: kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/AppConfig.java (hunk 2 above)
Comment (woitech) [reply kodilla-good-patterns/src/main/java/com/kodilla/good/patterns/food2door/AppConfig.java]: Nobody will add new company inside a dummy parseeConfigFile(String configFile). Companies' data are beeing written in a config file and simultaneously an adequate implementation of their offer and order services are coded. I simply skip a support of the config file.  So then I  provided only a part of the solution. So I must correct it.

**Review by aszecowka — APPROVED** [review]

**Pull request merged**